Sources/BSON/BSONBinary.swift

```diff
@@ -0,0 +1,206 @@
+    public let data: ByteBuffer
+
+    /// The binary subtype for this data.
+    public let subtype: Subtype
+
+    /// Subtypes for BSON Binary values.
+    public struct Subtype: Equatable, Hashable, RawRepresentable {
+        // swiftlint:disable force_unwrapping
+        /// Generic binary subtype
+        public static let generic = Subtype(rawValue: 0x00)!
+        /// A function
+        public static let function = Subtype(rawValue: 0x01)!
+        /// Binary (old)
+        public static let binaryDeprecated = Subtype(rawValue: 0x02)!
+        /// UUID (old)
+        public static let uuidDeprecated = Subtype(rawValue: 0x03)!
+        /// UUID (RFC 4122)
+        public static let uuid = Subtype(rawValue: 0x04)!
+        /// MD5
+        public static let md5 = Subtype(rawValue: 0x05)!
+        /// Encrypted BSON value
+        public static let encryptedValue = Subtype(rawValue: 0x06)!
+        // swiftlint:enable force_unwrapping
+
+        /// Subtype indicator value
+        public let rawValue: UInt8
+
+        /// Initializes a `Subtype` with a custom value.
+        /// Returns nil if rawValue within reserved range [0x07, 0x80).
+        public init?(rawValue: UInt8) {
+            guard !(rawValue > 0x06 && rawValue < 0x80) else {
+                return nil
+            }
+            self.rawValue = rawValue
+        }
+
+        /// Initializes a `Subtype` with a custom value. This value must be in the range 0x80-0xFF.
+        /// - Throws:
+        ///   - `BSONError.InvalidArgumentError` if value passed is outside of the range 0x80-0xFF
+        public static func userDefined(_ value: Int) throws -> Subtype {
+            guard let byteValue = UInt8(exactly: value) else {
+                throw BSONError.InvalidArgumentError(message: "Cannot represent \(value) as UInt8")
+            }
+            guard byteValue >= 0x80 else {
+                throw BSONError.InvalidArgumentError(
+                    message: "userDefined value must be greater than or equal to 0x80 got \(byteValue)"
+                )
+            }
+            guard let subtype = Subtype(rawValue: byteValue) else {
+                throw BSONError.InvalidArgumentError(message: "Cannot represent \(byteValue) as Subtype")
+            }
+            return subtype
+        }
+    }
+
+    /// Initializes a `BSONBinary` instance from a `UUID`.
+    /// - Throws:
+    ///   - `BSONError.InvalidArgumentError` if a `BSONBinary` cannot be constructed from this UUID.
+    public init(from uuid: UUID) throws {
+        let uuidt = uuid.uuid
+
+        let uuidData = Data([
+            uuidt.0, uuidt.1, uuidt.2, uuidt.3,
+            uuidt.4, uuidt.5, uuidt.6, uuidt.7,
+            uuidt.8, uuidt.9, uuidt.10, uuidt.11,
+            uuidt.12, uuidt.13, uuidt.14, uuidt.15
+        ])
+
+        try self.init(data: uuidData, subtype: BSONBinary.Subtype.uuid)
+    }
+
+    /// Initializes a `BSONBinary` instance from a `Data` object and a `UInt8` subtype.
```

Comment:
this docstring should say `Subtype` instead of `UInt8`. Perhaps it should call out that this always creates a copy of the data regardless of CoW behavior.